Remove diameter

joaander opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should be based on *trunk-patch*. -->
<!-- Backwards compatible new features should be based on *trunk-minor*. -->
<!-- Incompatible API changes should be based on *trunk-major*. -->

## Description

<!-- Describe your changes in detail. -->
* Remove diameter shifting from neighbor list.
* Remove `alpha` parameter from Langevin and related integration methods.
* Remove diameter from potential evaluators.
* `write.GSD` now requires opt-in to write diameters.

## Motivation and context

<!--- Why is this change required? What problem does it solve? -->
Starting in v3.0.0, particle diameters no longer influence the simulation. The removed code is no longer used. Some plugins abuse `diameter` as an arbitrary per-particle quantity, so it remains in `ParticleData` for now. See #1266 for more discussion on this topic.

<!-- Replace ??? with the issue number that this pull request resolves. -->
Resolves #1266

## How has this been tested?

Existing unit tests.

<!--- Please describe in detail how you tested your changes. -->

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Change log

<!-- Propose a change log entry. -->
```
Changed:

* ``hoomd.write.GSD`` does not write ``particles/diameter`` by default.

Removed:

* ``alpha`` parameter and attribute in `Langevin`, `BD`, and `OverdampedViscous` integration methods.
* ``needsDiameter`` and ``setDiameter`` API in C++ potential evaluators.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

joaander committed:
Remove particle diameters in DLVO test.

DLVO does not use per-particle diameters.

joaander committed:
Remove neighbor list diameter shift tests.

This functionality will be removed from the neighbor list.

joaander committed:
Remove diameter shift warning from PPPM.

joaander committed:
Remove diameter from potentials.

joaander committed:
Remove alpha from Langevin and derived methods.

joaander committed:
Remove diameter shift in neighbor list.

joaander committed:
Require user to opt-into writing diameters in GSD files.

joaander committed:
Document diameter changes in migration guide.

b-butler commented on hoomd/md/PotentialBondGPU.cuh:
```diff
@@ -145,22 +141,13 @@ __global__ void gpu_compute_bond_forces_kernel(Scalar4* d_force,
         q = __ldg(d_charge + idx);
         }
     else
-        q += 0; // shut up compiler warning
```
Haha :joy:
someone was having a bad day.

b-butler commented on hoomd/md/methods/rattle.py:
```diff
@@ -391,44 +360,28 @@ class Brownian(MethodRATTLE):
-                 manifold_constraint,
-                 tolerance=0.000001,
-                 alpha=None):
+    def __init__(self, filter, kT, manifold_constraint, tolerance=0.000001):
```
```suggestion
    def __init__(self, filter, kT, manifold_constraint, tolerance=1e-6):
```
Not necessary but easier to read.

b-butler commented on hoomd/md/methods/rattle.py:
```diff
@@ -511,41 +458,24 @@ class OverdampedViscous(MethodRATTLE):
-                 manifold_constraint,
-                 tolerance=0.000001,
-                 alpha=None):
+    def __init__(self, filter, manifold_constraint, tolerance=0.000001):
```
```suggestion
    def __init__(self, filter, manifold_constraint, tolerance=1e-6):
```

b-butler commented on hoomd/md/NeighborListGPUBinned.cu:
```diff
@@ -217,7 +214,7 @@ __global__ void gpu_compute_nlist_binned_kernel(unsigned int* d_nlist,
             if (!use_index)
                 cur_tdb = d_cell_tdb[cli(cur_offset, neigh_cell)];
             else
-                cur_tdb = make_scalar4(postype_j.w, d_diameter[j], __int_as_scalar(d_body[j]), 0);
```
Is this 0 just to prevent more refactoring?

joaander replied:
Yes, the cell list (optionally) repacks the particle type, diameter and body ("tdb") in a memory-efficient layout for `NeighborListBinned`.

Now that I look at it, `NeighborListBinned` is the only consumer of this data structure, so I should refactor `CellList` to store only type and body on this branch, removing this 0.

b-butler commented on hoomd/write/gsd.py:
```diff
@@ -141,6 +141,9 @@ class GSD(Writer):
         truncate (bool): When `True`, truncate the file and write a new frame 0
             each time this operation triggers.
         dynamic (list[str]): Quantity categories to save in every frame.
+        write_diameter (bool): When `False`, do not write
```
Should this be part of dynamic? It sort of fulfills the same role except being only a single property.

joaander replied:
I made it separate because it isn't the same as dynamic. If `diameter` was outside the dynamic list, then it would still be written on frame 0. I would like to avoid writing it at all unless the user opts in.

joaander replied:
Refactoring complete. As a bonus, the more efficient memory accesses boost performance by a few % in the LJ hoomd benchmark: 3030 TPS to 3130.

b-butler commented on hoomd/metal/EAMForceCompute.cc:
```diff
@@ -355,9 +355,6 @@ void EAMForceCompute::computeForces(uint64_t timestep)
     Scalar r_cut_sq = m_r_cut * m_r_cut;
 
-    // sum up the number of forces calculated
-    int64_t n_calc = 0;
```
Why the removals here in metal?

joaander replied:
The latest Apple clang generates warnings for the unused `n_calc`. This change is unrelated and could be cherry-picked into another pull request if necessary.

joaander replied:
I also found a few more unused `benchmark` functions and removed them. The removal can be cherry-picked to another branch if needed.

b-butler replied:
No need to worry about that then. Just wondering what prompted the seemingly unrelated change.

cbkerr replied:
👍 I wondered about benchmark removal too

b-butler reviewed (approved):
Looks good. I left some comments but nothing blocking.

joaander committed:
Apply suggestions from code review

Co-authored-by: Brandon Butler <butlerbr@umich.edu>

joaander committed:
Removed unused benchmark() method.

This change is not related to the diameter pull request and can be
cherry-picked to a separate branch if needed.

joaander committed:
Replace TDB with TypeBody

Removes diameter from the cell list arrays.

b-butler commented:
Its fine. I approved because I didn't think any of my comments should stop the PR, plus we don't squash our commits in PRs so having some cleanup is not the same as in `signac`.

cbkerr commented on hoomd/CellList.h:
```diff
@@ -313,9 +313,9 @@ class PYBIND11_EXPORT CellList : public Compute
         }
 
     //! Get the cell list containing t,d,b
```
```suggestion
    //! Get the cell list containing t,b
```

cbkerr commented on hoomd/md/EvaluatorPairExpandedMie.h:
```diff
@@ -29,39 +29,6 @@ namespace hoomd
 namespace md
     {
 //! Class for evaluating the Expanded Mie pair potential
-/*! <b>General Overview</b>
```
Are c doc strings removed to maintain less?

joaander replied:
Yes, this is another cleanup unrelated to the diameter removal except that I noticed it while going through the code.

I removed out of date comments in Evaluator files that either 1) duplicated comments from the prototypical example (`EvaluatorPairLJ` for pair potentials) or 2) referenced out of date functional forms (e.g. Expanded Mie with particle diameters). For the latter, the functional form in the Python documentation should be the authoritative reference.

cbkerr commented on hoomd/md/EvaluatorPairTWF.h:
```diff
@@ -35,69 +35,6 @@ namespace hoomd
-/*! <b>General Overview</b>
-
-    EvaluatorPairTWF is a low level computation class that computes the TWF pair
-    potential V(r). As the standard MD potential, it also serves as a well
-    documented example of how to write additional pair potentials. "Standard"
-    pair potentials in hoomd are all handled via the template class
```
Is there a documented example of this elsewhere?

cbkerr reviewed (approved):
The migration guide looks good. Does it need a changelog entry?

joaander committed:
Apply suggestion from code review.

Co-authored-by: Corwin Kerr <cbkerr@umich.edu>

joaander commented:
> The migration guide looks good. Does it need a changelog entry?

I can add a blanket change log entry stating that the migration guide now contains details in migrating scripts from v3 to v4.
